### PR TITLE
[WIP] QMC5883 Magnetometer Driver

### PR DESCRIPTION
**Pull request by dlwalter**
Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary.

**Test data / coverage**
[sensors status](https://i.imgur.com/HLhuCXK.png)
[listener sensor_mag -i 0](https://i.imgur.com/Rb8qqYv.png)
[tests sensors](https://i.imgur.com/2dot8il.png)

Magnetometer plots - QMC5883 External mag is sensor_mag_1
[X axis plot comparison](https://i.imgur.com/4FxjZep.png)
[Y axis plot comparison](https://i.imgur.com/3b78WPQ.png)
[Z axis plot comparison](https://i.imgur.com/xAO3V6z.png)

**Describe problem solved by the proposed pull request**
A driver for the QMC5883 is required as many manufacturers have moved away from the now obsolete HMC5883.

**Describe your preferred solution**
The QMC5883 driver is implemented with basic functionality.  Currently runs at a fixed range (+/- 2Gauss), oversampling (512) and bitrate (200hz).  The next steps are to implement functionality to change these configuration options and implement a self-test.   Doesn't work with current stable branch dues to the use of MAGIOSELFTEST, will result in preflight check errors.

**Describe possible alternatives**
Write another driver

**Additional context**

### CODE REVIEW

File: boards/px4/fmu-v2/init/rc.board [hunk 1]
```diff
@@ -43,6 +43,7 @@ fi
 
 # External I2C bus
 hmc5883 -C -T -X start
+qmc5883 -C -T -X start
```
Comment (dlwalter) [inline boards/px4/fmu-v2/init/rc.board]: This command fails to start the driver.  Still have to manually start at the command line.

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 2]
```diff
@@ -0,0 +1,1640 @@
...
+	}
+
+	ret = OK;
+
+out:
+	perf_end(_sample_perf);
+	return ret;
+}
+
+int QMC5883::calibrate(struct file *filp, unsigned enable)
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: Need to confirm that QMC5883::calibrate does anything

Comment (dlwalter) [reply src/drivers/magnetometer/qmc5883/qmc5883.cpp]: This function does run when the "-C" option is given for the HMC5883.  @dagar does the calibrated values just get discarded?

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 3]
```diff
@@ -0,0 +1,1640 @@
...
+	ret = read_reg(QMC5883_ADDR_CONTROL_1, range_bits_in);
+
+	if (OK != ret) {
+		perf_count(_comms_errors);
+		return;
+	}
+
+	if (range_bits_in != _range_bits << 4) {
+		perf_count(_range_errors);
+		ret = write_reg(QMC5883_ADDR_CONTROL_1, (_range_bits << 4));
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: @dagar I think this is why it was going STALE.  The first 2 bits of the control address (0x09) are the data rate which are getting written over here and putting it back into standby mode.

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 4]
```diff
@@ -0,0 +1,1544 @@
...
+	_sensor_ok = true;
+out:
+	return ret;
+}
+
+/**
+   check that the configuration register has the right value. This is
+   done periodically to cope with I2C bus noise causing the
+   configuration of the compass to change.
+ */
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: removed get and check range and just hardcoded the +/-2G range

File: boards/px4/fmu-v3/init/rc.board [hunk 5]
```diff
@@ -137,6 +139,11 @@ else
 	lsm303d start
 fi
 
+# External I2C bus magnetometer
+qmc5883 -X start
```
Comment (dlwalter) [inline boards/px4/fmu-v3/init/rc.board]: this command still doesn't work.  I have to manually start the sensor.

File: boards/px4/fmu-v2/init/rc.board [hunk 6]
```diff
@@ -43,6 +43,7 @@ fi
 
 # External I2C bus
 hmc5883 -C -T -X start
+qmc5883 -X start
```
Comment (dlwalter) [inline boards/px4/fmu-v2/init/rc.board]: Removed the calibration option on this command.

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 7]
```diff
@@ -0,0 +1,1179 @@
...
+
+	/**
+	 * check the sensor configuration.
+	 *
+	 * checks that the config of the sensor is correctly set, to
+	 * cope with communication errors causing the configuration to
+	 * change
+	 */
+	void 			check_conf(void);
+
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: Removed the calibration function - QMC5883 doesn't have a self-test function with an induced magnetic field like the HMC5983 has.

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 8]
```diff
@@ -0,0 +1,1179 @@
...
+/* Control Register 2 */
+#define QMC5883_INT_ENB 			(1 << 0)
+#define QMC5883_ROL_PNT 			(1 << 6)
+#define QMC5883_SOFT_RESET 			(1 << 7)
+
+/* Set Register */
+#define QMC5883_SET_DEFAULT 			(1 << 0)
+
+#define QMC5883_TEMP_OFFSET 50 /* deg celsius */
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: Temp offset is not calibrated on sensor at the factory so this is a pretty rough estimate to get ballpark sensor temp.  All temp calibration is happening on chip by default.

File: boards/px4/fmu-v2/init/rc.board (hunk 1 above)
Comment (dlwalter) [reply boards/px4/fmu-v2/init/rc.board]: Fixed this issue in the probe() command for i2c.  Have to read 0x00 once first before you can read the ID registers reliably.

**Comment by jinger26** [comment]
The new Holybro Pixhawk 4 GPS is also pending this driver. cc @bkueng

**Comment by bys1123** [comment]
@jinger26 Thanks, actually it's not Pixhawk4‘s GPS module, old GPS modules which used HMC5983 now replaced by QMC5883.

**Comment by Antiheavy** [comment]
We are considering this part for an upcoming board design.  Will be interested to evaluate once this PR goes through.

File: src/drivers/magnetometer/qmc5883/qmc5883_i2c.cpp [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	read(0x00, &data[0], 1);
+	read(0x00, &data[0], 1);
+
+	_retries = 10;
+
+	unsigned count = _retries;
+	bool read_valid = false;
+	bool id_valid = false;
+
+	while(count > 0){
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883_i2c.cpp]: I added a number of retries to probe for in cases of poor wiring or i2c interference.  This has helped with the driver not starting on boot.

File: boards/px4/fmu-v2/init/rc.board (hunk 1 above)
Comment (dlwalter) [reply boards/px4/fmu-v2/init/rc.board]: I also added a while loop to check a number of retries.

File: src/drivers/magnetometer/qmc5883/qmc5883.cpp [hunk 10]
```diff
@@ -0,0 +1,1180 @@
...
+#include <lib/conversion/rotation.h>
+
+#include "qmc5883.h"
+
+/*
+ * QMC5883 internal constants and data structures.
+ */
+
+/* Max measurement rate is 200Hz */
+#define QMC5883_CONVERSION_INTERVAL	(1000000 / 150)	/* microseconds */
```
Comment (dlwalter) [inline src/drivers/magnetometer/qmc5883/qmc5883.cpp]: @dagar I made the denomator 150 to match the HMC driver, but I recently noticed that the HMC5883 driver is actually polling at 50hz (measuring on a logic analyzer).  I think this is due to the HMC5883:calibrate function here [setting the polling rate to 50hz](https://github.com/PX4/Firmware/blob/9782aecc73b45ae72ce01e507b1535c43f3780b6/src/drivers/magnetometer/hmc5883/hmc5883.cpp#L1041) and never getting corrected.

**Comment by dagar** [comment]
I think this could use a bit more work, but it looks safe to merge. I've updated with latest master and disabled the autostart for now. Let's test it a bit people enabling by default. Nice work!

**Pull request merged**